src/jit/gentree.cpp

```diff
@@ -6586,115 +6799,165 @@ GenTreePtr Compiler::gtNewAssignNode(GenTreePtr dst, GenTreePtr src)
 }
 
 // Creates a new Obj node.
-GenTreeObj* Compiler::gtNewObjNode(CORINFO_CLASS_HANDLE structHnd, GenTree* addr)
+GenTree* Compiler::gtNewObjNode(CORINFO_CLASS_HANDLE structHnd, GenTree* addr)
 {
     var_types nodeType = impNormStructType(structHnd);
     assert(varTypeIsStruct(nodeType));
-    GenTreeObj* objNode = new (this, GT_OBJ) GenTreeObj(nodeType, addr, structHnd);
+    unsigned size = info.compCompHnd->getClassSize(structHnd);
+
+    // It would be convenient to set the GC info at this time, but we don't actually require
+    // it unless this is going to be a destination.
+    if (!varTypeIsStruct(nodeType))
+    {
+        if ((addr->gtOper == GT_ADDR) && (addr->gtGetOp1()->TypeGet() == nodeType))
+        {
+            return addr->gtGetOp1();
+        }
+        else
+        {
+            return gtNewOperNode(GT_IND, nodeType, addr);
+        }
+    }
+    GenTreeBlk* newBlkOrObjNode = new (this, GT_OBJ) GenTreeObj(nodeType, addr, structHnd, size);
+
     // An Obj is not a global reference, if it is known to be a local struct.
-    GenTreeLclVarCommon* lclNode = addr->IsLocalAddrExpr();
-    if ((lclNode != nullptr) && !lvaIsImplicitByRefLocal(lclNode->gtLclNum))
+    if ((addr->gtFlags & GTF_GLOB_REF) == 0)
     {
-        objNode->gtFlags &= ~GTF_GLOB_REF;
+        GenTreeLclVarCommon* lclNode = addr->IsLocalAddrExpr();
+        if ((lclNode != nullptr) && !lvaIsImplicitByRefLocal(lclNode->gtLclNum))
+        {
+            newBlkOrObjNode->gtFlags &= ~GTF_GLOB_REF;
+        }
     }
-    return objNode;
+    return newBlkOrObjNode;
 }
 
-// Creates a new CpObj node.
-// Parameters (exactly the same as MSIL CpObj):
+//------------------------------------------------------------------------
+// gtSetObjGcInfo: Set the GC info on an object node
 //
-//  dst        - The target to copy the struct to
-//  src        - The source to copy the struct from
-//  structHnd  - A class token that represents the type of object being copied. May be null
-//               if FEATURE_SIMD is enabled and the source has a SIMD type.
-//  isVolatile - Is this marked as volatile memory?
-GenTreeBlkOp* Compiler::gtNewCpObjNode(GenTreePtr dst, GenTreePtr src, CORINFO_CLASS_HANDLE structHnd, bool isVolatile)
-{
-    size_t    size       = 0;
-    unsigned  slots      = 0;
-    unsigned  gcPtrCount = 0;
-    BYTE*     gcPtrs     = nullptr;
-    var_types type       = TYP_STRUCT;
-
-    GenTreePtr hndOrSize = nullptr;
+// Arguments:
+//    objNode - The object node of interest
 
-    GenTreeBlkOp* result = nullptr;
+void Compiler::gtSetObjGcInfo(GenTreeObj* objNode)
+{
+    CORINFO_CLASS_HANDLE structHnd  = objNode->gtClass;
+    var_types            nodeType   = impNormStructType(structHnd);
```

Comment:
Note that impNormStructType() is going to all getClassSize() and getClassGCLayout() EE methods.  These could already be called while creating GenTreeObj node and gets called again.  Since nodeType set on GenTreeObj is that of normalized type, cant' we use the type of objNode here?